### PR TITLE
Add time and block advancement

### PR DESCRIPTION
**Pull request by p-offtermatt**
## Description

Closes: #1006 

CometMock now offers time advancement, block advancement and causing downtime via CometMock-specific RPC endpoints.

See https://github.com/informalsystems/CometMock/commit/4c231434b1b292bb657f4a6ea99744130f0c55c1 for changes on the CometMock side.

Changes should mostly be self-explanatory.

New baseline comparison with the time and block advancement is that CometBFT takes 4min50s for the short-happy-path,
CometMock takes 1min40s for it.

You can confirm this by running `go run ./tests/e2e --verbose --use-cometmock --use-gorelayer --short-happy-path --verbose` for CometMock and `go run ./tests/e2e --verbose --use-gorelayer --short-happy-path --verbose` for non-CometMock

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] Provided a link to the relevant issue or specification
- [x] Reviewed "Files changed" and left comments if necessary 
- [x] Confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] Confirmed all author checklist items have been addressed
- [ ] Confirmed that this PR does not change production code

### CODE REVIEW

File: tests/e2e/state.go [hunk 1]
```diff
@@ -199,6 +199,16 @@ func (tr TestRun) getBlockHeight(chain chainID) uint {
 }
 
 func (tr TestRun) waitBlocks(chain chainID, blocks uint, timeout time.Duration) {
```
Comment (jtremback) [inline tests/e2e/state.go:201]: May want to rename this to newBlocks or something since it doesn't always wait

**Review by jtremback — APPROVED** [review]
LGTM

File: tests/e2e/actions.go [hunk 2]
```diff
@@ -171,6 +172,17 @@ func (tr TestRun) startChain(
 		chain:     action.chain,
 		validator: action.validators[0].id,
 	}, verbose)
+
+	// store the fact that we started the chain
+	if tr.runningChains == nil {
```
Comment (smarshall-spitzbart) [inline tests/e2e/actions.go:177]: nit imo a more common pattern for this initialization would be a constructor for `TestRun` that initializes the map

File: tests/e2e/actions.go [hunk 3]
```diff
@@ -1035,7 +1047,7 @@ func (tr TestRun) transferChannelComplete(
 	executeCommand(chanOpenConfirmCmd, "transferChanOpenConfirm")
 }
 
-func executeCommand(cmd *exec.Cmd, cmdName string) {
+func executeCommandWithVerbosity(cmd *exec.Cmd, cmdName string, verbose *bool) {
```
Comment (smarshall-spitzbart) [inline tests/e2e/actions.go:1050]: Nit for basic types
```suggestion
func executeCommandWithVerbosity(cmd *exec.Cmd, cmdName string, verbose bool) {
```

File: tests/e2e/actions.go [hunk 4]
```diff
@@ -1063,6 +1075,10 @@ func executeCommand(cmd *exec.Cmd, cmdName string) {
 	}
 }
 
+func executeCommand(cmd *exec.Cmd, cmdName string) {
```
Comment (smarshall-spitzbart) [inline tests/e2e/actions.go:1078]:
```suggestion
// Executes a command with verbosity specified by CLI flag
func executeCommand(cmd *exec.Cmd, cmdName string) {
```

**Review by smarshall-spitzbart — APPROVED** [review]
Nice! LGTM

File: tests/e2e/actions.go (hunk 2 above)
Comment (p-offtermatt) [reply tests/e2e/actions.go:177]: I put this into a seperate TestRun.initialize function. My problem with the constructor approach is that go doesn't support named arguments for funcs but for struct creation named arguments work. The calls where the TestRuns are constructed become huge blocks of 10 arguments which are only identifiable by their ordering if they go through a constructor, which I found worse than this solution

File: tests/e2e/state.go (hunk 1 above)
Comment (p-offtermatt) [reply tests/e2e/state.go:212]: I think it's better like this - even with CometMock, the name still gets at what it does, namely after this is done executing a number of blocks will have passed

**Comment by p-offtermatt** [comment]
Thanks for the reviews, great suggestions!

**Pull request merged**